lib/tests/test_working_copy.rs

```diff
@@ -210,73 +212,120 @@ fn test_checkout_file_transitions(use_git: bool) {
+            diff_snapshot,
+            DiffSummary {
+                added: vec![
+                    "GitSubmodule_Conflict",
+                    "GitSubmodule_Executable",
+                    "GitSubmodule_ExecutableNormalContent",
+                    "GitSubmodule_Normal",
+                    "GitSubmodule_Symlink",
+                ]
+                .iter()
+                .map(|str| RepoPath::from_internal_string(str))
+                .collect_vec(),
+                modified: vec![],
+                removed: vec![],
+            }
+        );
+    } else {
+        assert_eq!(new_tree_id, right_tree_id);
+    }
+
+    for (left_kind, right_kind, path) in &files {
         let wc_path = workspace_root.join(path.to_internal_file_string());
         let maybe_metadata = wc_path.symlink_metadata();
-        match right_kind {
-            Kind::Missing => {
-                assert!(maybe_metadata.is_err(), "{:?} should not exist", path);
-            }
-            Kind::Normal => {
-                assert!(maybe_metadata.is_ok(), "{:?} should exist", path);
-                let metadata = maybe_metadata.unwrap();
-                assert!(metadata.is_file(), "{:?} should be a file", path);
-                #[cfg(unix)]
-                assert_eq!(
-                    metadata.permissions().mode() & 0o111,
-                    0,
-                    "{:?} should not be executable",
-                    path
-                );
-            }
-            Kind::Executable | Kind::ExecutableNormalContent => {
-                assert!(maybe_metadata.is_ok(), "{:?} should exist", path);
-                let metadata = maybe_metadata.unwrap();
-                assert!(metadata.is_file(), "{:?} should be a file", path);
-                #[cfg(unix)]
-                assert_ne!(
-                    metadata.permissions().mode() & 0o111,
-                    0,
-                    "{:?} should be executable",
-                    path
-                );
-            }
-            Kind::Conflict => {
-                assert!(maybe_metadata.is_ok(), "{:?} should exist", path);
-                let metadata = maybe_metadata.unwrap();
-                assert!(metadata.is_file(), "{:?} should be a file", path);
-                #[cfg(unix)]
-                assert_eq!(
-                    metadata.permissions().mode() & 0o111,
-                    0,
-                    "{:?} should not be executable",
-                    path
-                );
-            }
-            Kind::Symlink => {
-                assert!(maybe_metadata.is_ok(), "{:?} should exist", path);
-                let metadata = maybe_metadata.unwrap();
-                assert!(
-                    metadata.file_type().is_symlink(),
-                    "{:?} should be a symlink",
-                    path
-                );
-            }
-            Kind::Tree => {
+        match (left_kind, right_kind) {
+            (Kind::GitSubmodule, Kind::Tree) => {
```

Comment:
I'm also unsure why updating to a `tree` works, maybe because it's just ignored and never actually written?